app/scripts/components/common/mapbox/layers/styles.tsx

```diff
@@ -0,0 +1,122 @@
+import { AnyLayer, AnySourceImpl, Style } from 'mapbox-gl';
+import React, { createContext, useCallback, useEffect, useState } from 'react';
+
+export type ExtendedLayer = AnyLayer & {
+  metadata?: {
+    layerOrderPosition?: LayerOrderPosition;
+    [key: string]: any;
+  };
+};
+export interface GeneratorParams {
+  generatorId: string;
+  layers: ExtendedLayer[];
+  sources: Record<string, AnySourceImpl>;
+  metadata?: Record<string, unknown>;
+}
+
+interface StylesContextType {
+  updateStyle: (params: GeneratorParams) => void;
+}
+
+export const StylesContext = createContext<StylesContextType>({
+  updateStyle: (params: GeneratorParams) => {
+    return params;
+  }
+});
+
+export type LayerOrderPosition =
+  | 'basemap-background'
+  | 'raster'
+  | 'markers'
+  | 'vector'
+  | 'basemap-foreground';
+
+const LAYER_ORDER: LayerOrderPosition[] = [
+  'basemap-background',
+  'raster',
+  'vector',
+  'basemap-foreground',
+  'markers'
+];
+
+// Takes in a dictionary associating each generator id with a series of
+//  Mapbox layers and sources to be added to the final style. Outputs
+//  a style object directly usable by the map instance.
+const generateStyle = (stylesData: Record<string, GeneratorParams>) => {
+  let sources: Record<string, AnySourceImpl> = {};
+  let layers: ExtendedLayer[] = [];
+
+  Object.entries(stylesData).map(([generatorId, generatorParams]) => {
+    // TODO check duplicate source ids?
+    sources = {
+      ...sources,
+      ...generatorParams.sources
+    };
+
+    const layersWithMeta = [
+      ...generatorParams.layers.map((layer) => {
+        const metadata = layer.metadata ?? {};
+        metadata.generatorId = generatorId;
+        return { ...layer, metadata };
+      })
+    ];
+
+    layers = [...layers, ...layersWithMeta];
+  });
```

Comment:
To keep the current approach, I'd use a `foreach` instead of a `map`. Same outcome, but no return expectations.

To avoid reassigning variables, this could be a `reduce`.